httprule/httprule_test.go

```diff
@@ -0,0 +1,143 @@
+package httprule
+
+import (
+	"strings"
+	"testing"
+
+	"foxygo.at/protog/httprule/internal"
+	"github.com/stretchr/testify/require"
+	pb "google.golang.org/genproto/googleapis/api/annotations"
+	"google.golang.org/protobuf/proto"
+)
+
+func TestValidateHTTPRule(t *testing.T) {
+	r := &pb.HttpRule{
+		Pattern: &pb.HttpRule_Get{Get: "abc"},
+		Body:    "abc",
+	}
+	require.Error(t, ValidateHTTPRule(r))
+	r.Body = "*"
+	require.Error(t, ValidateHTTPRule(r))
+	r.Body = ""
+	require.NoError(t, ValidateHTTPRule(r))
+
+	r = &pb.HttpRule{
+		Pattern: &pb.HttpRule_Delete{Delete: "abc"},
+		Body:    "abc",
+	}
+	require.Error(t, ValidateHTTPRule(r))
+	r.Body = "*"
+	require.Error(t, ValidateHTTPRule(r))
+	r.Body = ""
+	require.NoError(t, ValidateHTTPRule(r))
+}
+
+func requireProtoEqual(t *testing.T, want, got proto.Message) {
+	t.Helper()
+	require.True(t, proto.Equal(want, got), "protos are not Equal")
+}
+
+func requireProtoNotEqual(t *testing.T, want, got proto.Message) {
+	t.Helper()
+	require.False(t, proto.Equal(want, got), "protos are Equal")
+}
+
+func TestParseProtoResponseMsg1(t *testing.T) {
+	rule := &pb.HttpRule{}
+	s := `{"field1": "val1"}`
+	got := &internal.TestMessage1{}
+	want := &internal.TestMessage1{Field1: "val1"}
+	err := ParseProtoResponse(rule, strings.NewReader(s), got)
+	require.NoError(t, err)
+	requireProtoEqual(t, want, got)
+
+	s = `{"field1": "XXXX"}`
+	require.NoError(t, ParseProtoResponse(rule, strings.NewReader(s), got))
+	requireProtoNotEqual(t, want, got)
+}
+
+func TestParseProtoResponseMsg2(t *testing.T) {
+	rule := &pb.HttpRule{}
+	s := `{"field1": "val1"}`
+	got := &internal.TestMessage2{}
+	want := &internal.TestMessage2{Field1: "val1"}
+	err := ParseProtoResponse(rule, strings.NewReader(s), got)
+	require.NoError(t, err)
+	requireProtoEqual(t, want, got)
+
+	s = `{}`
+	want = &internal.TestMessage2{}
+	require.NoError(t, ParseProtoResponse(rule, strings.NewReader(s), got))
+	requireProtoEqual(t, want, got)
+
+	s = `{"field2": 0}`
+	require.NoError(t, ParseProtoResponse(rule, strings.NewReader(s), got))
+	requireProtoEqual(t, want, got)
+
+	s = `{"field2": 3}`
+	want = &internal.TestMessage2{Field2: 3}
+	require.NoError(t, ParseProtoResponse(rule, strings.NewReader(s), got))
+	requireProtoEqual(t, want, got)
+
+	s = `{"field3Sub": {"subField": "abc"} }`
+	want = &internal.TestMessage2{Field3Sub: &internal.SubMessage{SubField: "abc"}}
+	require.NoError(t, ParseProtoResponse(rule, strings.NewReader(s), got))
+	requireProtoEqual(t, want, got)
+}
+
+func TestField(t *testing.T) {
+	sub := &internal.SubMessage{SubField: "abc"}
```

Comment:
can you put this directly into `m` instead of creating a var here, like you do below in `wantMsg`? The var does not seem to be used anywhere else.

Reply:
ok.